lib/engine/auto_router.rb

```diff
@@ -88,81 +133,328 @@ def compute(corporation, **opts)
+          else
+            puts "  ERROR: auto-router found unexpected number of path node edges #{node1.edges.size}. "\
+                 'Route combos may be be incorrect'
           end
+          index += 1
         end
+      end
+      bitfield
+    end
+
+    def check_and_set(bitfield, hexside_left, hexside_right, hexside_bits, bit_to_bitmask)
+      check_edge_and_set(bitfield, hexside_left, hexside_bits, bit_to_bitmask)
+      check_edge_and_set(bitfield, hexside_right, hexside_bits, bit_to_bitmask)
+    end
 
-        combos.compact!
-      rescue RuntimeError
-        puts 'Route timeout reach'
-        break
+    def check_edge_and_set(bitfield, hexside_edge, hexside_bits, bit_to_bitmask)
+      if hexside_bits.include?(hexside_edge)
+        set_bit(bitfield, hexside_bits[hexside_edge], bit_to_bitmask)
+      else
+        newbit = hexside_bits.length.positive? ? hexside_bits.map { |_hexside, bit| bit }.max + 1 : 0
+        hexside_bits[hexside_edge] = newbit
+        set_bit(bitfield, newbit, bit_to_bitmask)
       end
+    end
 
-      puts "Found #{possibilities.size} possible routes in: #{Time.now - now}"
+    # bitfield is an array of integers, can be expanded by this call if necessary
+    # bit is a bit number, 0 is lowest bit, 32 will jump to the next int in the array, and so on
+    def set_bit(bitfield, bit, bit_to_bitmask)
+      entry = (bit / 32).to_i
+      mask = bit_to_bitmask[bit & 31]
+      add_count = entry + 1 - bitfield.size
+      while add_count.positive?
+        bitfield << 0
+        add_count -= 1
+      end
+      bitfield[entry] |= mask
+    end
 
-      max_routes = possibilities.max_by do |routes|
-        routes.each { |route| route.routes = routes }
-        @game.routes_revenue(routes)
-      end || []
+    # does testroute conflict with any route in the combo?
+    def route_bitfield_conflicts(combo, testroute)
+      combo.each do |route|
+        # only test up to the shorter size, since bits beyond that obviously don't conflict
+        index = [route.bitfield.size, testroute.bitfield.size].min - 1
+        while index >= 0
+          return true if (route.bitfield[index] & testroute.bitfield[index]) != 0
 
-      max_routes.each { |route| route.routes = max_routes }
+          index -= 1
+        end
+      end
+      false
     end
+
+    # The js-in-Opal algorithm
+    def js_evaluate_combos(rb_sorted_routes, _route_timeout)
+      rb_possibilities = []
+      possibilities_count = 0
+      conflicts = 0
+      limit = rb_sorted_routes.map(&:size).reduce(&:*) # rubocop:disable Lint/UselessAssignment
+      now = Time.now
+
+      if RUBY_ENGINE == 'opal'
+        %x(
+        let possibilities = []
+        let combos = [];
+        let counter = 0;
+        let max_revenue = 0;
+
+        // marshal Opal objects to js for faster/easier access
+        js_sorted_routes = [];
+        Opal.send(rb_sorted_routes, 'each', [], function(rb_routes)
```

Comment:
why do you have rb_sorted_routes and js_sorted_routes? you should probably only have one, please clean this up